pkg/controller/controller.go

```diff
@@ -561,6 +561,24 @@ func (p *csiProvisioner) ProvisionExt(options controller.ProvisionOptions) (*v1.
 		return nil, controller.ProvisioningInBackground, capErr
 	}
 
+	if options.PVC.Spec.DataSource != nil && (rc.clone || rc.snapshot) {
+		contentSource := rep.GetVolume().ContentSource
+		if contentSource == nil {
+			sourceErr := fmt.Errorf("volume content source missing")
+			delReq := &csi.DeleteVolumeRequest{
+				VolumeId: rep.GetVolume().GetVolumeId(),
+			}
+			delReq.Secrets = provisionerCredentials
+			ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+			defer cancel()
+			_, err := p.csiClient.DeleteVolume(ctx, delReq)
+			if err != nil {
+				sourceErr = fmt.Errorf("%v. cleanup of volume %s failed, volume is orphaned: %v", sourceErr, pvName, err)
```

Comment:
do we need to try to delete volume a few times if it fails? Another option is still create PV object, but indicate the source is missing. This will require API change and need more discussion.

Reply:
What does provisioner currently do?

Reply:
Currently the provisioner silently creates an empty volume if the content source is missing because there is no check.  That's why this PR tries to add a check there.

I think it is not useful to create an empty volume when user wants to create a volume from snapshot.  Typically user wants to do this because the original volume is corrupted so they want to restore the volume.  An empty volume won't solve their problems.  The reason that the content source is missing is most likely because the driver didn't implement "create volume from snapshot" so it just goes through "create a new volume" code path.

Reply:
The concern here is how to deal with the orphaned volumes

Reply:
Similar delete volume logic is on line 556 in this routine.  We should handle orphaned volumes consistently.  Retries may help.

Reply:
Let's retry Delete a few times. That's what pv controller does

Reply:
@zhucan can you add retries for volume clean up here?  If delete volume fails, retry a few times.

Reply:
copy that, I will do it @xing-yang  @msau42